### PR TITLE
ARROW-3762: [C++] manage ChunkedArrayBuilder capacity explicitly

### PR DESCRIPTION
**Pull request by bkietz**
ChunkedArrayBuilder should never request BinaryBuilder reserve space for more strings than it can hold (`kListMaximumElements`); with this change it will instead of begin a new chunk

### CODE REVIEW

**Review by wesm — COMMENTED** [review]
This looks okay, but I think we need to add some unit tests (that aren't disabled) to exercise the various code paths in `ChunkedBinaryBuilder::Reserve` to assert that they do what we expect. Otherwise if we refactor later we're going to be playing whackamole

File: cpp/src/arrow/array/builder_binary.h [hunk 1]
```diff
@@ -347,15 +357,16 @@ class ARROW_EXPORT ChunkedBinaryBuilder {
...
 
   virtual Status Finish(ArrayVector* out);
 
  protected:
   Status NextChunk();
 
   int64_t max_chunk_size_;
-  int64_t chunk_data_size_;
+  int64_t chunk_data_size_ = 0;
+  int64_t extra_capacity_ = 0;
```
Comment (pitrou) [inline cpp/src/arrow/array/builder_binary.h]: What is this? Can you add a comment?

**Review by pitrou — COMMENTED** [review]
Looks like a reasonable improvement to me. Unfortunately it'll be difficult to come up with a test that doesn't blow 16+GB of memory.

**Comment by bkietz** [comment]
I'll add some tests to src/arrow/array-binary-test.cc

File: cpp/src/arrow/array/builder_binary.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/array/builder_binary.h]: will do

**Comment by bkietz** [comment]
@pitrou @wesm This new test is pretty minimal but still takes ~4 minutes to pass on my machine

**Comment by wesm** [comment]
Hm. Can we make the 2GB limit something that's configurable rather than hard-coded? In theory that's the purpose of the max chunk size. We can't have such a long-running test in CI

**Comment by bkietz** [comment]
`max_chunk_size_` is a restriction on the character data (string_array.buffer[2]); the test I added appends 2 ** 31 *empty* strings so `chunk_data_size_` never rises from 0.

We could certainly add another to apply to the # of strings in the string_array, or alternatively use `max_chunk_size_ / sizeof(int32_t)` as that limit

**Comment by wesm** [comment]
Oh okay, sorry for my confusion. Can we make the number of elements per chunk configurable, then? I can see a use case for having no more than 64K elements per chunk, no matter what their size. Then this is much easier to test

**Comment by bkietz** [comment]
Will do

**Review by wesm — APPROVED** [review]
+1, thanks @bkietz! Will merge once CI has run

**Comment by wesm** [comment]
Merging, the failure is the cloud.r-project.org DNS issue

**Pull request closed without merging**